### ISSUE TITLE
sql: disable direct access to the system.settings table

### ISSUE THREAD

**Issue by knz** [cockroachdb/cockroach#15250]
Right now users can set cluster settings using the `set cluster setting ...` statement, and view the settings using various ways.

Meanwhile  users with sufficient privileges (including root) can write directly to `system.setting`, including inserting non-sensical values in there.  

It seems to me we want to disallow direct access to this table even for the root user. Thoughts?

cc @petermattis @bdarnell

**Comment by knz** [comment]
Also the RFC was pretty explicit:

> The `settings` table will not have write privileges at the SQL layer, but will
instead be read-only, and only readable by the root user, thus forcing the use
of the `SET` statement, ensuring validation and allowing for changes to be in a
`settings_history` table (e.g. with a`(name,time)` key and the old value).

The fact the code is doing something else is ... strange

**Comment by dt** [comment]
Yep, as you found, this is discussed in the RFC, implementation of which is in-progress.

The table was added with its privilege validation that will accept either r/w or r/o to facilitate this change (h/t to @benesch for that).

**Comment by knz** [comment]
I looked at this and saw it is hard to not have the write permission if we want a `set cluster setting` statement to do the right thing; unless, that is, we had a super-super-user which the clients cannot use but which `set cluster setting` could use under the hood.

**Comment by dt** [comment]
When I looked at this initially, I was thinking we'll want something similar to what we do for privilege checks when servicing view queries, e.g. the `SET` takes full responsibility for statement privilege check, then opts its internal executor out of doing further checks.

**Comment by knz** [comment]
How do you do "opt the internal executor out"?

**Comment by dt** [comment]
same way views queries currently do it: flag on internal planner.

**Comment by dt** [comment]
pushing to 1.2 -- i haven't seen this used in practice / customer clusters long enough yet to be confident we won't want to tell someone to, as root, run a `update system.settings...` or something in the course of debugging. it's root-only, and a system table, so i don't think this is quite time-critical (and the hardcoded allowed permissions allow us to do it via migration whenever we want).
